internals/overlord/logstate/clienterr/clienterr.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+/*
+Package clienterr contains common error types which are recognised by the log
+gatherer. logstate.logClient implementations should return these error types
+to communicate with the gatherer.
+
+Errors in this package should be pattern-matched using errors.As:
+
+		err := client.Flush(ctx)
+		backoff := &clienterr.Backoff{}
+		if errors.As(err, &backoff) {
+	      ...
+		}
+*/
+package clienterr
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+)
+
+// Backoff should be returned if the server indicates we are sending too many
+// requests (e.g. an HTTP 429 response).
+type Backoff struct {
+	RetryAfter *time.Time
+}
+
+func (e *Backoff) Error() string {
+	errStr := "too many requests"
+	if e.RetryAfter != nil {
+		errStr += ", retry after " + e.RetryAfter.String()
+	}
+	return errStr
+}
+
+// ErrorResponse represents an HTTP error response from the server
+// (4xx or 5xx).
+type ErrorResponse struct {
```

Comment:
I don't really understand why this type was moved here or why it needs to be here. The `clienterr` package, as I understand it, is only for errors that are part of the logging client contract. Whereas this isn't used anywhere and doesn't need to be part of that contract (it would only apply to HTTP-based clients, for one thing).

Happy to discuss if you disagree or I'm missing something.

Reply:
This falls more under the category of "common logic". There's nothing Loki-specific about it (only HTTP-specific, as you note), and it seemed cleaner to me to put it in the `clienterr` package.

It might be helpful for the gatherer to recognise this error type, but clients should prefer using a more specific type when appropriate (e.g. `Backoff` for a 429). This is the generic "default" case.

Do you think I should move it inside the `logstate/loki` package?

Reply:
I kind of agree with your reasoning in this case, @barrettj12

Reply:
I think you're making a good point @barrettj12